src/ports/ORMPUpgradeableAndRetryablePort.sol

```diff
@@ -0,0 +1,82 @@
+// This file is part of Darwinia.
+// Copyright (C) 2018-2023 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.17;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "ORMP/src/user/UpgradeableApplication.sol";
+import "./ORMPPort.sol";
+
+contract ORMPUpgradeableAndRetryablePort is ORMPPort, UpgradeableApplication, ReentrancyGuard {
+    /// @dev msgHash => isDispathedInPort.
+    mapping(bytes32 => bool) public dones;
+
+    event MessageDispatchedInPort(bytes32 indexed msgHash);
+    event ClearFailedMessage(bytes32 indexed msgHash);
+
+    constructor(address dao, address ormp, string memory name) ORMPPort(dao, ormp, name) UpgradeableApplication(ormp) {}
+
+    function protocol() public view override(Application, UpgradeableApplication) returns (address) {
+        return super.protocol();
+    }
+
+    function setORMP(address ormp) external onlyOwner {
+        _setORMP(ormp);
+    }
+
+    function retryFailedMessage(Message calldata message) external payable nonReentrant {
+        bytes32 msgHash = _checkMessage(message);
+        (, address fromDapp, address toDapp, bytes memory payload) =
+            abi.decode(message.encoded, (bytes4, address, address, bytes));
+        _recv(message.fromChainId, fromDapp, toDapp, payload);
+        _markDone(msgHash);
+    }
+
+    function _checkMessage(Message calldata message) internal view returns (bytes32 msgHash) {
+        msgHash = hash(message);
+        require(IORMP(protocol()).dones(msgHash) == true, "!done");
+        require(LOCAL_CHAINID() == message.toChainId, "!toChainId");
+        require(address(this) == message.to, "!to");
+        uint256 fromChainId = message.fromChainId;
+        require(message.from == _checkedFromPort(fromChainId), "!xAuth");
+    }
```

Comment:
Here, it checks the msgHash has been done in `IORMP(protocol())`, but didn't check that it has not been done in this contract `dones[msgHash]`, this may cause one message been recv multiple times.